### PR TITLE
Updating SuperLU package file with ability to build version 4.3.

### PR DESCRIPTION
**Pull request by jrood-nrel**
Updating SuperLU package file with ability to build version 4.3. (before they added cmake) with external blas or internal blas, and added capability to build the latest version 5.2.1 (with cmake) with internal blas.

### CODE REVIEW

**Comment by jrood-nrel** [comment]
Ah, I forgot the flake8 check. I will fix this when I get a chance. I was also reading the discussion about -fPIC in another pull request I may need to address in a more portable fashion.

File: var/spack/repos/builtin/packages/superlu/package.py [hunk 1]
```diff
@@ -34,21 +35,86 @@ class Superlu(Package):
...
+            'CDEFS      = -DAdd_'
+        ])
+
+        # Write configuration options to include file
+        with open('make.inc', 'w') as inc:
+            for option in config:
+                inc.write('{0}\n'.format(option))
+
+    def install(self, spec, prefix):
+        if spec.satisfies('@5.2.1:'):
```
Comment (davydden) [inline var/spack/repos/builtin/packages/superlu/package.py]: i think the recommended way it to shift this into a decorator 
```
@when('@5.2.1:')
def install(self, spec, prefix):
```
and then another `install` below.

File: var/spack/repos/builtin/packages/superlu/package.py [hunk 2]
```diff
@@ -34,21 +35,86 @@ class Superlu(Package):
     url      = "http://crd-legacy.lbl.gov/~xiaoye/SuperLU/superlu_5.2.1.tar.gz"
 
     version('5.2.1', '3a1a9bff20cb06b7d97c46d337504447')
+    version('4.3', 'b72c6309f25e9660133007b82621ba7c')
 
-    depends_on('blas')
+    variant('blas', default=True, description='Build with external BLAS library')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/superlu/package.py]: there is no consensus about this, but my opinion is that not packages should ever used internal BLAS. Imagine what happens if you have a graph of packages which are used to build something, all build as shared libraries, and each use it's own BLAS, some 32bit, some 64bit, some with threading, some without, etc.

So, i am not in favour of adding this variant.

File: var/spack/repos/builtin/packages/superlu/package.py [hunk 3]
```diff
@@ -34,21 +35,86 @@ class Superlu(Package):
...
+                   '-Denable_blaslib=OFF',
+                   '-DBLAS_blas_LIBRARY={0}'.format(spec['blas'].blas_libs.joined())
+               ])
+           else:
+               cmake_args.extend([
+                   '-Denable_blaslib=ON'
+               ])
+           cmake_args.extend(std_cmake_args)
+           with working_dir('spack-build', create=True):
+               cmake('..', *cmake_args)
```
Comment (davydden) [inline var/spack/repos/builtin/packages/superlu/package.py]: looks like `cmake` build-dependency was/is missing.

File: var/spack/repos/builtin/packages/superlu/package.py (hunk 2 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/superlu/package.py]: see https://github.com/LLNL/spack/issues/2392

Comment (alalazo) [inline var/spack/repos/builtin/packages/superlu/package.py]: I share @davydden opinion on having packages sneaking dependencies in, in particular for libraries. If we start to do this, there's no way spack can ensure the correctness of a build.

File: var/spack/repos/builtin/packages/superlu/package.py (hunk 1 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/superlu/package.py]: As you start with a conditional, the recommended way is to use a decorator for multi-methods. Write the default `install` without decorator as usual, the following ones should read instead:
```python
@when('<condition>')
def install(self, spec, prefix):
    ....
```

File: var/spack/repos/builtin/packages/superlu/package.py [hunk 4]
```diff
@@ -34,21 +35,86 @@ class Superlu(Package):
...
+               make()
+               make('install')
+        else:
+           #Pre-cmake installation method
+           self.configure(spec)
+           if '+blas' not in spec:
+               make('blaslib',parallel=False)
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/superlu/package.py]: Se the comment on the variant above.

**Comment by jrood-nrel** [comment]
The requested changes make sense. I will work on them. The reason I am adding SuperLU 4.3 is that Trilinos has not been updated to use the SuperLU 5.x API.

**Comment by davydden** [comment]
> The reason I am adding SuperLU 4.3 is that Trilinos has not been updated to use the SuperLU 5.x API.

right, that's why only superlu-dist is currently used in Trilinos.

**Pull request merged**